File: src/mimetype_parser.erl
-module(mimetype_parser).

-export([parse/1]).
-export([parse/2]).

parse(Src) ->
  parse(Src, 1).

parse(Src, LineNum) when is_integer(LineNum) ->
  case mimetype_parser_lexer:string(to_string(Src), LineNum) of
    {ok, Tokens, _} ->
      mimetype_parser_parser:parse(Tokens);
    Error ->
      Error
  end;
parse(Src, #{line := LineNum}) ->
  parse(Src, LineNum);
parse(Src, _) ->
  parse(Src, 1).

to_string(Bin) when is_list(Bin) ->
  Bin;
to_string(Bin) when is_binary(Bin) ->
  case unicode:characters_to_list(Bin) of
    L when is_list(L) ->
      L;
    _ ->
      binary_to_list(Bin)
  end.

File: lib/mimetype_parser.ex
defmodule MimetypeParser do
  defmodule Exception do
    defexception [:message]
  end
  def parse(string), do: :mimetype_parser.parse(string)
  def parse(string, opts), do: :mimetype_parser.parse(string, opts)

  def parse!(string), do: parse!(string, 1)
  def parse!(string, opts) do
    case parse(string, opts) do
      {:ok, types} ->
        types
      {:error, {_, :mimetype_parser_lexer, error}, _} ->
        raise Exception, message: to_string(:mimetype_parser_lexer.format_error(error))
      {:error, {_, :mimetype_parser_parser, parser_error}} ->
        raise Exception, message: to_string(parser_error)
    end
  end
end
